sa/storage-authority.go

```diff
@@ -673,6 +673,125 @@ func (ssa *SQLStorageAuthority) FinalizeAuthorization(authz core.Authorization)
 	return
 }
 
+const collectAuthIDs = `SELECT id FROM %s
+WHERE identifier = :ident AND
+status != :invalid AND
+status != :revoked AND
+expires > :now`
+
+// RevokeAuthorizationsByDomain invalidates all pending or finalized authorization
+// and any associated challenges for a single domain
+func (ssa *SQLStorageAuthority) RevokeAuthorizationsByDomain(ident core.AcmeIdentifier) (int64, int64, error) {
+	invalidStatus, revokedStatus := string(core.StatusInvalid), string(core.StatusRevoked)
+	identifier, err := json.Marshal(ident)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	tx, err := ssa.dbMap.Begin()
+	if err != nil {
+		return 0, 0, err
+	}
+
+	// collect authorization IDs before beginning revocations
+	now := ssa.clk.Now()
+	var allIDs []string
+	_, err = tx.Select(
+		&allIDs,
+		fmt.Sprintf(collectAuthIDs, "authz"),
```

Comment:
Instead of a constant string, this might be better as a func?

The API would then be typesafe insted of this tricky biz with getting the map key names right.